core/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java

```diff
@@ -555,15 +556,29 @@ Integer parse(XContentParser parser, boolean coerce) throws IOException {
 
             @Override
             Query termQuery(String field, Object value) {
-                int v = parse(value);
+                if (hasDecimalPart(value)) {
+                    return Queries.newMatchNoDocsQuery("Value [" + value + "] has a decimal part");
+                }
+                int v = parse(value, true);
                 return IntPoint.newExactQuery(field, v);
             }
 
             @Override
             Query termsQuery(String field, List<Object> values) {
-                int[] v = new int[values.size()];
-                for (int i = 0; i < values.size(); ++i) {
-                    v[i] = parse(values.get(i));
+                final List<Object> nonDecimalValues = new ArrayList<>();
+
+                values.forEach(value -> {
+                    if (!hasDecimalPart(value)) {
+                        nonDecimalValues.add(value);
+                    }
+                });
+                if (nonDecimalValues.isEmpty()) {
+                    return Queries.newMatchNoDocsQuery("All values have a decimal part");
+                }
+
+                int[] v = new int[nonDecimalValues.size()];
+                for (int i = 0; i < nonDecimalValues.size(); ++i) {
+                    v[i] = parse(nonDecimalValues.get(i), true);
```

Comment:
Maybe we could avoid generating intermediate garbage because of boxed objects by directly creating the primitive array, eg. something like:

```java
int[] v = new int[values.size()];
int upTo = 0;
for (Object value : values) {
  if (hasDecimalPart(value) == false) {
    v[upTo++] = parse(value, true);
  }
}
if (upTo != v.length) {
  v = Arrays.copyOf(v, upTo);
}
```